File: ApacKiosk/Services/CameraCapture.cs
using System;
using System.Drawing;
using System.Drawing.Imaging;
using System.IO;
using System.Threading;
using ApacKiosk.Database;
using ApacKiosk.Utils;
using AForge.Video.DirectShow;

namespace ApacKiosk.Services
{
    public class CameraCapture
    {
        private readonly ConfigManager _config;
        private readonly DatabaseManager _db;
        private Thread _thread;
        private volatile bool _isRunning;
        private int? _currentUserId;
        private VideoCaptureDevice _videoDevice;

        public CameraCapture(ConfigManager config, DatabaseManager db)
        {
            _config = config;
            _db = db;
        }

        public void SetCurrentUser(int? userId)
        {
            _currentUserId = userId;
        }

        public void Start()
        {
            if (_isRunning) return;
            _isRunning = true;
            _thread = new Thread(CaptureLoop)
            {
                IsBackground = true,
                Name = "CameraCapture"
            };
            _thread.Start();
        }

        public void Stop()
        {
            _isRunning = false;
            _thread?.Join(3000);
        }

        private void CaptureLoop()
        {
            while (_isRunning)
            {
                try
                {
                    if (_config.CameraEnabled && _currentUserId.HasValue)
                    {
                        CaptureNow();
                    }
                }
                catch (Exception ex)
                {
                    _db.InsertLog(_currentUserId, "camera_error", null, ex.Message);
                    Thread.Sleep(30000);
                }

                var sleepTime = Math.Max(_config.CameraIntervalSec, 30) * 1000;
                var slept = 0;
                while (slept < sleepTime && _isRunning)
                {
                    Thread.Sleep(100);
                    slept += 100;
                }
            }
        }

        public void CaptureNow()
        {
            var devices = new FilterInfoCollection(FilterCategory.VideoInputDevice);
            if (devices.Count == 0)
            {
                _db.InsertLog(_currentUserId, "camera_error", null, "Câmera não encontrada");
                return;
            }

            if (_videoDevice == null || !_videoDevice.IsRunning)
            {
                _videoDevice = new VideoCaptureDevice(devices[0].MonikerString);
                _videoDevice.NewFrame += OnNewFrame;
                _videoDevice.Start();
                Thread.Sleep(1000);
            }
        }

        private void OnNewFrame(object sender, AForge.Video.NewFrameEventArgs eventArgs)
        {
            try
            {
                var userLabel = _currentUserId?.ToString() ?? "system";
                var timestamp = DateTime.Now.ToString("yyyyMMdd_HHmmss");
                var dir = _config.CameraPath;
                Directory.CreateDirectory(dir);

                var fileName = $"camera_{timestamp}_{userLabel}.jpg";
                var filePath = Path.Combine(dir, fileName);

                int quality;
                switch (_config.CameraQuality)
                {
                    case "High": quality = 90; break;
                    case "Low": quality = 40; break;
                    default: quality = 70; break;
                }

                var encoder = ImageCodecInfo.GetImageEncoders();
                var jpegEncoder = Array.Find(encoder, e => e.MimeType == "image/jpeg");
                var encoderParams = new EncoderParameters(1);
                encoderParams.Param[0] = new EncoderParameter(System.Drawing.Imaging.Encoder.Quality, quality);

                using (var frame = (Bitmap)eventArgs.Frame.Clone())
                {
                    frame.Save(filePath, jpegEncoder, encoderParams);
                }
                encoderParams.Dispose();

                _db.InsertLog(_currentUserId, "camera", filePath, null);

                if (_videoDevice != null && _videoDevice.IsRunning)
                {
                    _videoDevice.SignalToStop();
                    _videoDevice.NewFrame -= OnNewFrame;
                    _videoDevice = null;
                }
            }
            catch (Exception ex)
            {
                _db.InsertLog(_currentUserId, "camera_error", null, ex.Message);
                try
                {
                    if (_videoDevice != null && _videoDevice.IsRunning)
                    {
                        _videoDevice.SignalToStop();
                        _videoDevice.NewFrame -= OnNewFrame;
                        _videoDevice = null;
                    }
                }
                catch { }
            }
        }

        public static string[] GetCameraDevices()
        {
            try
            {
                var devices = new FilterInfoCollection(FilterCategory.VideoInputDevice);
                var names = new string[devices.Count];
                for (int i = 0; i < devices.Count; i++)
                    names[i] = devices[i].Name;
                return names;
            }
            catch
            {
                return Array.Empty<string>();
            }
        }
    }
}

File: apac_kiosk/monitoring/camera_capture.py
from __future__ import annotations

import os
import threading
import time
from datetime import datetime

import cv2

from ..database import db


class CameraCapture:
    def __init__(self):
        self._running = False
        self._thread: threading.Thread | None = None
        self._user_id: int | None = None
        self._username: str = ""
        self._cap: cv2.VideoCapture | None = None

    def start(self, user_id: int, username: str):
        self._user_id = user_id
        self._username = username
        device_id = int(db.get_config("camera_device", "0"))
        self._cap = cv2.VideoCapture(device_id)
        if not self._cap.isOpened():
            db.insert_log(None, "camera_error", description="Câmera não encontrada ou indisponível")
            self._cap = None
            return
        self._running = True
        self._thread = threading.Thread(target=self._capture_loop, daemon=True)
        self._thread.start()

    def stop(self):
        self._running = False
        if self._cap:
            self._cap.release()
            self._cap = None

    def _capture_loop(self):
        while self._running:
            try:
                interval = int(db.get_config("camera_interval", "120"))
                self._capture_once()
            except Exception:
                pass
            time.sleep(interval)

    def _capture_once(self):
        if self._cap is None:
            return
        try:
            ret, frame = self._cap.read()
            if not ret:
                db.insert_log(None, "camera_error", description="Falha ao capturar frame da câmera")
                return

            quality = db.get_config("camera_quality", "medium")
            quality_map = {"high": 95, "medium": 70, "low": 40}
            jpg_quality = quality_map.get(quality, 70)

            timestamp = datetime.now().strftime("%Y%m%d_%H%M%S")
            filename = f"camera_{timestamp}_{self._username}.jpg"
            folder = db.get_config("camera_path", os.path.join(os.path.dirname(db.DB_PATH), "camera"))
            os.makedirs(folder, exist_ok=True)
            filepath = os.path.join(folder, filename)

            cv2.imwrite(filepath, frame, [cv2.IMWRITE_JPEG_QUALITY, jpg_quality])
            db.insert_log(
                self._user_id, "camera", file_path=filepath,
                description="Captura de câmera"
            )
        except Exception:
            pass

    @staticmethod
    def list_cameras() -> list[tuple[int, str]]:
        cameras = []
        for i in range(10):
            cap = cv2.VideoCapture(i)
            if cap.isOpened():
                cameras.append((i, f"Câmera {i}"))
                cap.release()
        return cameras
